tools/releaseBuild/azureDevOps/templates/windows-packaging.yml

```diff
@@ -261,7 +263,81 @@ jobs:
         Write-Host "Uploading $packagePath"
         Write-Host "##vso[artifact.upload containerfolder=signed;artifactname=signed]$packagePath"
       }
-    displayName: Upload packages
+    displayName: Upload unsigned packages
+
+  - ${{ if and(ne(variables['BuildConfiguration'],'minSize'), startsWith(variables['Architecture'], 'x')) }}:
+    - template: EsrpSign.yml@ComplianceRepo
+      parameters:
+          buildOutputPath: $(System.ArtifactsDirectory)\pkgSigned
+          signOutputPath: $(Build.StagingDirectory)\signedPackages
+          certificateId: "CP-230012"
+          pattern: |
+            **\*.msi
+          useMinimatch: true
+          shouldSign: $(SHOULD_SIGN)
+          displayName: Sign MSI
+
+    - powershell: |
+        Get-ChildItem '$(System.ArtifactsDirectory)\signedPackages' | ForEach-Object {
+          $packagePath = $_.FullName
+          Write-Host "Uploading $packagePath"
+          Write-Host "##vso[artifact.upload containerfolder=finalResults;artifactname=finalResults]$packagePath"
+        }
+      displayName: Upload signed MSI to finalResults
+
+    - task: AzureFileCopy@4
+      displayName: 'upload signed msi to Azure - ${{ parameters.architecture }}'
+      inputs:
+        SourcePath: '$(Build.StagingDirectory)\signedPackages\PowerShell-$(version)-win-${{ parameters.architecture }}.msi'
+        azureSubscription: '$(AzureFileCopySubscription)'
+        Destination: AzureBlob
+        storage: '$(StorageAccount)'
+        ContainerName: '$(AzureVersion)'
+        resourceGroup: '$(StorageResourceGroup)'
+
+    - pwsh: |
+        cd $(PowerShellRoot)
+        Import-Module $(PowerShellRoot)/build.psm1 -Force
+        Import-Module $(PowerShellRoot)/tools/packaging -Force
+
+        $msiPath = '$(Build.StagingDirectory)\signedPackages\PowerShell-$(version)-win-${{ parameters.architecture }}.msi'
+
+        New-ExePackage -ProductVersion '$(version)' -MsiLocationPath $msiPath -ProductTargetArchitecture ${{ parameters.architecture }}
+        $exePath = Get-ChildItem '.\PowerShell-*.exe' | Select-Object -First 1 -ExpandProperty fullname
+        $enginePath = Join-Path -Path '$(System.ArtifactsDirectory)\unsignedEngine' -ChildPath engine.exe
+        Dismount-ExePackageEngine -ExePath $exePath -EnginePath $enginePath
```

Comment:
I'm mostly not clear on the `Mount`/`Dismount-ExePackageEngine` uses here:

- Why is it `Mount`/`Dismount`? What does it leave on the filesystem?
- Why do we dismount in one task and then mount later in another?
- Are we depending on filesystem state between tasks?

Reply:
It's more like an archive.... so perhaps... `Expand`/`Compress`

Reply:
That makes some sense. What is exe package engine? What does each command do? Can we add comments that contain the answers to those questions and ideally the questions above?

Reply:
The engine of the exe is the installer enignine

Reply:
Ok having read a bit about [insignia](https://wixtoolset.org/documentation/manual/v3/overview/insignia.html), here are my recommendations:

- Rename the functions to `Disconnect-`/`Connect-ExePackageEngine`
- Add a comment in the function definitions like `Needed to detach the WIX Burn engine from the exe for signing` and `Needed to reattach the WIX Burn engine to the exe after signing`
- Add a comment in the YAML-embedded script where we detach saying `Detach the packaging engine so we can sign the exe`
- Add a similar comment when we re-attach saying `Re-attach the packaging engine to the now-signed exe so we can create the MSI`

It's not clear from this file why we need to re-attach Burn at the end though. We seem to do the following:

- sign the exe
- re-attach Burn to the exe
- upload it (but the display name says *unsigned*??)

I don't follow this sequence.